python/ray/serve/backend_state_manager.py

```diff
@@ -882,11 +948,19 @@ def _scale_backend_replicas(
             ReplicaState.SHOULD_START, ReplicaState.STARTING_OR_UPDATING,
             ReplicaState.RUNNING
         ])
+        failed_to_start_replicas = self._replicas[backend_tag].count(
+            states=[ReplicaState.FAILED_TO_START])
 
-        delta_replicas = target_replicas - current_replicas
+        delta_replicas = (
+            target_replicas - current_replicas - failed_to_start_replicas)
         if delta_replicas == 0:
             return False
 
+        if (failed_to_start_replicas == target_replicas
+                and target_replicas != 0):
+            logger.error("Deployment reconfiguration failed. ")
+            return False
```

Comment:
I'm still kind of confused about this `FAILED_TO_START` state -- why do we need it? It seems like all we need to do here is just count the number of replica failures that occur, we don't need to keep them around in this state.

Reply:
I'm treating it as a terminal state for replica start up for accurate replica state tracking. 

> It seems like all we need to do here is just count the number of replica failures that occur

That's exactly what we want to track, but the implementation matters. My initial attempt was to simply put them at stopped / should stop state, for non-transient failures we will think we never reached desirable # of replicas thus try to spin up replicas forever. Just counting raw # of retries is also not correct since it could vary on each replica.

So essentially what we need is be able to count # of replicas failed to start with no retires left, but also don't confuse them with the rest so new update() cycle won't try to move or loose track on them